src/Microsoft.ML.Tokenizers/Model/BertTokenizer.cs

```diff
@@ -762,54 +762,68 @@ private static BertTokenizer Create(
 
             options.Normalizer ??= options.ApplyBasicTokenization ? new BertNormalizer(options.LowerCaseBeforeTokenization, options.IndividuallyTokenizeCjk, options.RemoveNonSpacingMarks) : null;
 
+            Dictionary<string, int>? specialTokensDict = null;
             if (options.SplitOnSpecialTokens)
             {
                 bool lowerCase = options.ApplyBasicTokenization && options.LowerCaseBeforeTokenization;
                 if (options.SpecialTokens is not null)
                 {
                     if (lowerCase)
                     {
-                        Dictionary<string, int> dic = options.SpecialTokens.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
-                        options.SpecialTokens = dic;
-
+                        specialTokensDict = [];
                         foreach (var kvp in options.SpecialTokens)
                         {
                             if (!vocab.TryGetValue(new StringSpanOrdinalKey(kvp.Key), out int id) || id != kvp.Value)
                             {
                                 throw new ArgumentException($"The special token '{kvp.Key}' is not in the vocabulary or assigned id value {id} different than the value {kvp.Value} in the special tokens.");
                             }
 
-                            // Ensure that the special tokens are lowercased.
-                            dic[kvp.Key.ToLowerInvariant()] = kvp.Value;
+                            // Add the special token into our dictionary, normalizing it, and adding it into the
+                            // main vocab, if needed. 
+                            AddSpecialToken(vocab, specialTokensDict, kvp.Key, true);
                         }
                     }
+                    else
+                    {
+                        specialTokensDict = options.SpecialTokens?.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+                    }
```

Comment:
We don't have to do it this way allocating more dictionary. what we can do is:

- Define:
IReadOnlyDictionary<string, int>? specialTokensDict = options.SpecialTokens;

- change `specialTokensDict = [];` to soemthing like:

```
Dictionary<string, int> tempSpecialTokens = [];
pecialTokensDict = tempSpecialTokens;
```

- Then pass `tempSpecialTokens` to AddSpecialToken.

You'll need to same idea in the else part when the special tokens not provided in options.